### ISSUE TITLE
Upgrade to VS Version 17.4.0 Preview 2.0 and I can't drop Chart on form

### ISSUE THREAD

**Issue by paul1956** [kirsan31/winforms-datavisualization#4]
![image](https://user-images.githubusercontent.com/12789369/190839293-91ff5927-09d0-46d5-9867-f7fa005e58b5.png)

**Comment by kirsan31** [comment]
Nice....
Unfortunately, I can't install V.S. preview, and even if I could, it wouldn't help :( We have nice situation here...
- No designer sdk docs.
- Designer sdk still in preview.
- Preview of VS.

I will try to ask a question on this topic, but there is little chance of an answer, because still no response to the [previous one](https://github.com/dotnet/winforms/discussions/7073#:~:text=is%20defiantly%20needed.-,When%20reference%20a%20control%20by%20project%20reference%20and%20adding%20it%20from%20toolbar%20to%20the%20form%2C%20the%20exception%20appears%20(when%20reference%20a%20control%20by%20dll%20reference%20%2D%20all%20work%20good).,-exception)...

**Comment by paul1956** [comment]
Just an FYI the preview installs on a separate hive and runs parallel to your production version, I understand your doing it will not fix the issue.

**Comment by KlausLoeffelmann** [comment]
Can you try to attach a second version of VS both to the primary instance of VS and to the DesignToolsServer and Debug into it to see what exactly happens?

**Comment by paul1956** [comment]
@KlausLoeffelmann I have VS attached to both of the processes and neither hit anything. Just this message pops up and if you hit OK nothing is changes in the Solution.
![image](https://user-images.githubusercontent.com/12789369/191139928-c1d31985-f520-4da7-b64b-6af61c7cd6f8.png)



 If you open Form1 of https://github.com/paul1956/CareLink/tree/DataVisualizationDesigner in designer, no Charts show up in Preview 2 where I could mostly design in Preview 1. In Preview 1, if you got too deep the properties in Scrollbar for example below can't be click on or exited.

```
Me.HomeTabChart.ChartAreas(NameOf(ChartArea)).AxisX.ScrollBar
```

![image](https://user-images.githubusercontent.com/12789369/191139130-d7e95640-7c4a-410f-917a-e3c6911153e0.png)

The big black space on the left and behind and 100% are white Charts and Form1.Designer.vb has the correct add.

```
Me.SplitContainer3.Panel1.Controls.Add(Me.HomeTabChart)
Me.SplitContainer3.Panel2.Controls.Add(Me.HomeTabTimeInRangeChart)
```

**Comment by KlausLoeffelmann** [comment]
Did you change the exception filters to break on all exceptions?

![image](https://user-images.githubusercontent.com/9663150/191295170-243c3255-9b26-47f5-9a05-141cc8af40e4.png)

**Comment by paul1956** [comment]
@KlausLoeffelmann that helped
```C#
        public Assembly LoadFromAssemblyName(AssemblyName assemblyName)
        {
            if (assemblyName == null)
                throw new ArgumentNullException(nameof(assemblyName));

            // Attempt to load the assembly, using the same ordering as static load, in the current load context.
            StackCrawlMark stackMark = StackCrawlMark.LookForMyCaller;
            return RuntimeAssembly.InternalLoad(assemblyName, ref stackMark, this);
        }

```
System.IO.FileNotFoundException
  HResult=0x80070002
  Message=Could not load file or assembly 'System.Windows.Forms.DataVisualization, Version=1.3.8296.30984, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

```

**Comment by paul1956** [comment]
@KlausLoeffelmann this version also deleted everything related to DataVisualiztion including all handler clauses in for and everything in designer file if you do any edits to the Form the Control was dropped on. This was fixed for the last several releases.

**Comment by paul1956** [comment]
@KlausLoeffelmann do you know if this is fixed in Preview 3?

**Comment by KlausLoeffelmann** [comment]
I know too little about the approach.
What kind of Control Designer does it use?
Does it have dedicated Type Editors?
If yes, does it have Server and Client-versions for the Designer, and generate a NuGet Package which allows it to be loaded in the context of both the DesignToolsServer and VS?

I recently posted a few links to resources to check for best practices. The most important one is the preliminary (still in the making) description of the SDK: https://github.com/KlausLoeffelmann/NetControlDesigners/pull/9.

**Comment by paul1956** [comment]
@KlausLoeffelmann I can't answer those questions hopefully @kirsan31 can, I am just the user of the control. I do know that the Chart control worked in preview 1 (it appeared in toolbox, could be dropped on from, it displayed correctly and could be edited) and in Preview 2, existing charts don't display in designer and if **anything** changes on the form even unrelated to the Chart control everything about the chart is deleted from the design file except the Public Event declaration line. That seems like a major issue we have not seen since .Net 5 with Custom Controls.

**Comment by kirsan31** [comment]
@KlausLoeffelmann 

> I know too little about the approach.
> What kind of Control Designer does it use?
> Does it have dedicated Type Editors?
> If yes, does it have Server and Client-versions for the Designer, and generate a NuGet Package which allows it to be loaded in the context of both the DesignToolsServer and VS?

It's **very basic** port. Added sdk, corrected namespaces, replaced `ArrayEditor` with `Microsoft.DotNet.DesignTools.Editors.CollectionEditor`, minor edits, commented out the rest. So no client-server designer of course.

**Comment by paul1956** [comment]
No improvement in Preview 5
![image](https://user-images.githubusercontent.com/12789369/198905070-2a767e81-6e2a-4007-bd19-4a386396cc26.png)

I am dropping the control on a TabPage or TableLayoutPanel.

**Comment by KlausLoeffelmann** [comment]
No. I debugged into that over the weekend, and it seems, the System.CodeDom assembly is loaded twice - once from the global installation, and once from the shadow cache. And then, on instantiation, the external Serializer uses the shadow cache one, while the Designer used the global one. Since their CodeBase is different, the types are considered differently. We're still investigating, what broke to end up with this.

We don't seem to have that problem, when the server-part of a multi targeted Control Designer comes from a NuGet. But I have yet to confirm that, although I am pretty sure, since the examples and templates also use base classes of the CodeDom Serializer, and they never explode. That could be a workaround, and it would make sense to go down that route anyway, since the Type Editors must be used in the context of VS, anyway.

**Comment by kirsan31** [comment]
@KlausLoeffelmann I think you are confusing this error with #5. You are talking about #5 bug and it's defiantly can be workaround by nuget (because it's 100% workaround by dll reference ).

And here we are talking about:

> System.IO.FileNotFoundException
> HResult=0x80070002
> Message=Could not load file or assembly 'System.Windows.Forms.DataVisualization, Version=1.3.8296.30984, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.

that was appeared in 17.4 preview 2.

Or is it something I don't understand?

**Comment by KlausLoeffelmann** [comment]
Hmm.
Yes, I am obviously confusing those, although the initial symptoms were the same, no?
Is that only when you drop it on a Tab Control, or is the drop target irrelevant?
Because when I workaround #5 I can drop just fine.

**Comment by kirsan31** [comment]
@KlausLoeffelmann 

> the initial symptoms were the same, no?

No, #5 was always here from first port (it was V.S. 17.0). And it was only with project reference (with dll reference all good). Naturally, it also works through Nuget.

And this (#4) is a new issue from 17.4 preview 2 only. And this is with dll reference - it basically can't load the dll (those. with a high degree of probability with nuget it will be the same). 

> Is that only when you drop it on a Tab Control, or is the drop target irrelevant?

I think with clean form the error will be the same... @paul1956 can you pls check it?

**Comment by paul1956** [comment]
I was able to create a new project and add the DataVisualization Project and then drop the control on a "form". But when I tried to do something with it (like delete it) I get below.
![image](https://user-images.githubusercontent.com/12789369/199124841-c117c905-7c0c-4826-8b5e-bfac7b44a0f8.png)
If I try to drop on a TableLayoutPanel I get the same error
![image](https://user-images.githubusercontent.com/12789369/199125130-f5bbeda3-9bf7-4add-87a5-55d1dcf8c346.png)

**Comment by kirsan31** [comment]
@paul1956 both errors above are clearly #5. When your drag control on new clean form the reference to the control will be added **by project**.  But we need to test **dll reference** scenario.  It's my bad I did not describe exactly what needs to be done - sorry :(

The simplest way is to open `DataViz.sln` in latest V.S. preview, rebuild and drag chart control from the toolbar on `Form1` in the `DesignerTest.csproj` (reference to System.Windows.Forms.DataVisualization.dll is already present in the project file).

**Comment by paul1956** [comment]
If you mean
```
<ProjectReference Include="..\..\..\winforms-datavisualization\src\System.Windows.Forms.DataVisualization\ChartWin.csproj" />
```
```
 <ItemGroup>
   <Reference Include="System.Windows.Forms.DataVisualization">
     <HintPath>..\..\..\winforms-datavisualization\src\System.Windows.Forms.DataVisualization\bin\$(Configuration)\$(TargetFramework)\System.Windows.Forms.DataVisualization.dll</HintPath>
   </Reference>

They are both there

 </ItemGroup>
```

**Comment by kirsan31** [comment]
@ oh oh oh my bad again  🤦‍♂️🤦‍♂️🤦‍♂️

Yes they are both there and dll ref is commented out:

https://github.com/kirsan31/winforms-datavisualization/blob/22e89b960fd9a66c068954032566f80007578e87/DesignerTest/DesignerTest.csproj#L11-L19

For test we need to remove / comment out proj ref and uncomment dll ref:

```xml
<!--<ItemGroup>
  <ProjectReference Include="..\src\System.Windows.Forms.DataVisualization\ChartWin.csproj" />
</ItemGroup>-->

<ItemGroup>
  <Reference Include="System.Windows.Forms.DataVisualization">
    <HintPath>..\src\System.Windows.Forms.DataVisualization\bin\Debug\net6.0-windows\System.Windows.Forms.DataVisualization.dll</HintPath>
  </Reference>
</ItemGroup>
```

So, open `DataViz.sln` in latest V.S. preview, edit `DesignerTest.csproj` like above, save, rebuild and drag chart control from the toolbar on Form1 in the `DesignerTest.csproj` (reference to System.Windows.Forms.DataVisualization.dll is already present in the project file).

**Comment by paul1956** [comment]
@kirsan31 same error.
![image](https://user-images.githubusercontent.com/12789369/199858402-7e9a8431-b9aa-4b75-9f0d-15d669dbc55a.png)
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-windows</TargetFramework>
    <Nullable>enable</Nullable>
    <UseWindowsForms>true</UseWindowsForms>
    <ImplicitUsings>enable</ImplicitUsings>
  </PropertyGroup>

  <!--<ItemGroup>
    <ProjectReference Include="..\src\System.Windows.Forms.DataVisualization\ChartWin.csproj" />
  </ItemGroup>-->

  <ItemGroup>
    <Reference Include="System.Windows.Forms.DataVisualization">
      <HintPath>..\src\System.Windows.Forms.DataVisualization\bin\Debug\net6.0-windows\System.Windows.Forms.DataVisualization.dll</HintPath>
    </Reference>
  </ItemGroup>
```

**Comment by kirsan31** [comment]
/cc @KlausLoeffelmann 
I unsewered to you questions [here](https://github.com/kirsan31/winforms-datavisualization/issues/4#issuecomment-1297647066).

In 17.4 we have #5 with proj reference (this bug exist in all VS versions), and #4 with dll reference (only 17.4 bug).
We need you help, because after 17.4 come out we will loss designer completely :(

**Comment by KlausLoeffelmann** [comment]
So...good news is, we are actively investigating. Not so good news is, it's not at all obvious right now why this is happening, but we're on it. It may take a while, though, so please stay tuned. It's planned to be addressed in 17.5, with a chance to be serviced in 17.4, depending on the extend and additional effects of the fixed. @MerrieMcGaw FYI.

**Comment by kirsan31** [comment]
Thank you @KlausLoeffelmann. Then we stay at 17.3 and hope for servicing in 17.4...

**Comment by KlausLoeffelmann** [comment]
Quick additional questions in that context: 

a) Can you have 17.3 and 17.4 on your test machine at the same time or does installing 17.4 also screwing up the scenario? 

b) Is the Target Framework relevant (or installing additional .NET versions like 7.0 on that machine)?

Also @shyam-gupta for FYI to this.

**Comment by kirsan31** [comment]
@KlausLoeffelmann 

> a) Can you have 17.3 and 17.4 on your test machine at the same time or does installing 17.4 also screwing up the scenario?
> b) Is the Target Framework relevant (or installing additional .NET versions like 7.0 on that machine)?

17.4 have released already, So I can't install it side by side with 17.3 :( Theoretically I can try with 17.5...
@paul1956 can you help here?

**Comment by paul1956** [comment]
I am running the preview released yesterday. But it broke ModVb so I might revert. What is the exact request?

**Comment by kirsan31** [comment]
> I am running the preview released yesterday. But it broke ModVb so I might revert. What is the exact request?

Can you answer on @KlausLoeffelmann [questions above](https://github.com/kirsan31/winforms-datavisualization/issues/4#issuecomment-1309146320)?

**Comment by paul1956** [comment]
I think you can only have 1 release version and 1 preview version.

**Comment by kirsan31** [comment]
@paul1956 I'll try to rephrase the question: after you installed 17.4 preview2 (in which the designer didn't work), did the designer still work in 17.3?

**Comment by paul1956** [comment]
You don't think you can have the 2 full releases installed at the same time. I used GitHub to roll back and then stopped using designer.

**Comment by kirsan31** [comment]
@paul1956 
![image](https://user-images.githubusercontent.com/17767561/201043034-3796e013-ad56-4bdf-ae1a-d55987a25dcf.png)

On September 17 (when you post this issue) and then until 17.4 released - does designer worked in 17.3?

**Comment by KlausLoeffelmann** [comment]
> I think you can only have 1 release version and 1 preview version.

You can have as many VS versions as you want. They all run in parallel just fine, you just need to give them different folder names.

**Comment by kirsan31** [comment]
@KlausLoeffelmann 
I still don't understand how I can put the 17.4 release next to the 17.3 release 🤷‍♂️ but it's not that so important.

**Only 17.3 installed, no .net7 installed.**
We have #5 and no #4 (this).

**17.3 and 17.5p1 installed, .net7 installed (with 17.5p1).**
In 17.3 we still have #5 and no #4 (this).
In 17.5p1 we still have #5 **and** #4 (this).

So on your questions:

> a) Can you have 17.3 and 17.4 on your test machine at the same time or does installing 17.4 also screwing up the scenario?

17.3 working fine in any way and 17.4p2+ - no.

> b) Is the Target Framework relevant (or installing additional .NET versions like 7.0 on that machine)?

Installing additional .NET versions doesn't break 17.3.

**Comment by KlausLoeffelmann** [comment]
OK, that's helpful - thanks!
Again @Shyam-Gupta, @MerrieMcGaw for FYI.

**Comment by kirsan31** [comment]
[Fixed](https://github.com/dotnet/winforms/issues/8327#issuecomment-1338163834) in V.S. 17.4.**3**.
